File: typescript/tests/client/heartbeat.test.ts
import { describe, expect, it } from 'bun:test';
import { createHeartbeat, sendOneHeartbeat } from '../../src/client/heartbeat.ts';
import { MemoryTokenStore } from '../../src/client/token-store.ts';
import { mockFetchJson, mockFetchNetworkError } from './_helpers.ts';

describe('sendOneHeartbeat', () => {
  it('posts the token from the store and clears grace on success', async () => {
    const store = new MemoryTokenStore();
    await store.write({ token: 'LIC1.x', graceStartSec: 999 });
    let seen: unknown = null;
    const fetchImpl: typeof fetch = async (_url, init) => {
      const bodyStr = typeof init?.body === 'string' ? init.body : '';
      seen = JSON.parse(bodyStr);
      return new Response(JSON.stringify({ success: true, data: {} }), {
        status: 200,
        headers: { 'content-type': 'application/json' },
      });
    };
    const ok = await sendOneHeartbeat({
      baseUrl: 'https://x',
      fetchImpl,
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
      nowSec: () => 1234,
    });
    expect(ok).toBe(true);
    // Server contract: handleHeartbeat reads `token` from the body.
    // Earlier client versions sent {license_key, fingerprint,
    // runtime_version, timestamp} which the server never read — see the
    // revocation-latency hardening commit.
    expect(seen).toEqual({ token: 'LIC1.x' });
    expect((await store.read()).graceStartSec).toBeNull();
  });

  it('is a no-op when the store has no token', async () => {
    const store = new MemoryTokenStore();
    let called = false;
    const fetchImpl: typeof fetch = async () => {
      called = true;
      return new Response('{}', { status: 200 });
    };
    const ok = await sendOneHeartbeat({
      baseUrl: 'https://x',
      fetchImpl,
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
    });
    expect(ok).toBe(true);
    expect(called).toBe(false); // no token → no network call
  });

  it('returns false on network failure (non-fatal, no throw)', async () => {
    const store = new MemoryTokenStore();
    await store.write({ token: 'LIC1.x', graceStartSec: null });
    let errSeen: Error | null = null;
    const ok = await sendOneHeartbeat({
      baseUrl: 'https://x',
      fetchImpl: mockFetchNetworkError(),
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
      onError: (e) => {
        errSeen = e;
      },
    });
    expect(ok).toBe(false);
    expect(errSeen).not.toBeNull();
  });

  it('clears the store on LicenseRevoked response', async () => {
    const store = new MemoryTokenStore();
    await store.write({ token: 'soon-to-be-cleared', graceStartSec: null });
    let errSeen: { code?: string } | null = null;
    const fetchImpl: typeof fetch = async () =>
      new Response(
        JSON.stringify({
          success: false,
          error: { code: 'LicenseRevoked', message: 'license is revoked' },
        }),
        { status: 403, headers: { 'content-type': 'application/json' } },
      );
    const ok = await sendOneHeartbeat({
      baseUrl: 'https://x',
      fetchImpl,
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
      onError: (e) => {
        errSeen = e as unknown as { code?: string };
      },
    });
    expect(ok).toBe(false);
    expect(errSeen?.code).toBe('LicenseRevoked');
    expect((await store.read()).token).toBeNull();
  });

  it('clears the store on LicenseSuspended response', async () => {
    const store = new MemoryTokenStore();
    await store.write({ token: 'soon-to-be-cleared', graceStartSec: null });
    const fetchImpl: typeof fetch = async () =>
      new Response(
        JSON.stringify({
          success: false,
          error: { code: 'LicenseSuspended', message: 'license is suspended' },
        }),
        { status: 403, headers: { 'content-type': 'application/json' } },
      );
    const ok = await sendOneHeartbeat({
      baseUrl: 'https://x',
      fetchImpl,
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
    });
    expect(ok).toBe(false);
    expect((await store.read()).token).toBeNull();
  });

  it('does NOT clear the store on transient errors (rate limit)', async () => {
    const store = new MemoryTokenStore();
    await store.write({ token: 'must-survive', graceStartSec: null });
    let errSeen: { code?: string } | null = null;
    const fetchImpl: typeof fetch = async () =>
      new Response(null, { status: 429, headers: { 'retry-after': '30' } });
    const ok = await sendOneHeartbeat({
      baseUrl: 'https://x',
      fetchImpl,
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
      onError: (e) => {
        errSeen = e as unknown as { code?: string };
      },
    });
    expect(ok).toBe(false);
    expect(errSeen?.code).toBe('RateLimited');
    expect((await store.read()).token).toBe('must-survive');
  });

  it('CAS-guards against a parallel refresh that wrote a fresh token', async () => {
    // Simulate the race: heartbeat reads token "T1", server returns
    // LicenseRevoked, but BEFORE the heartbeat can clear the store a
    // parallel refresh wrote token "T2". The CAS guard must detect the
    // change and skip the clear so T2 survives.
    const store = new MemoryTokenStore();
    await store.write({ token: 'T1', graceStartSec: null });
    let stage = 0;
    const fetchImpl: typeof fetch = async () => {
      // While the request is in-flight (between the heartbeat's first
      // store.read and its second store.read inside the catch handler),
      // simulate a parallel refresh by writing T2.
      if (stage === 0) {
        stage = 1;
        await store.write({ token: 'T2', graceStartSec: null });
      }
      return new Response(
        JSON.stringify({
          success: false,
          error: { code: 'LicenseRevoked', message: 'license is revoked' },
        }),
        { status: 403, headers: { 'content-type': 'application/json' } },
      );
    };
    await sendOneHeartbeat({
      baseUrl: 'https://x',
      fetchImpl,
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
    });
    // T2 must survive — CAS guard prevents clobbering.
    expect((await store.read()).token).toBe('T2');
  });
});

describe('createHeartbeat', () => {
  it('clamps interval below minimum to 60s', async () => {
    const store = new MemoryTokenStore();
    let warned: Error | null = null;
    const hb = createHeartbeat({
      baseUrl: 'https://x',
      fetchImpl: mockFetchJson(200, { success: true, data: {} }),
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
      intervalSec: 10,
      onError: (e) => {
        warned = e;
      },
    });
    expect(warned).not.toBeNull();
    expect((warned as unknown as Error).message).toContain('below minimum');
    hb.stop();
  });

  it('tickNow calls the issuer immediately', async () => {
    const store = new MemoryTokenStore();
    await store.write({ token: 'LIC1.x', graceStartSec: null });
    let calls = 0;
    const fetchImpl: typeof fetch = async () => {
      calls++;
      return new Response(JSON.stringify({ success: true, data: {} }), {
        status: 200,
        headers: { 'content-type': 'application/json' },
      });
    };
    const hb = createHeartbeat({
      baseUrl: 'https://x',
      fetchImpl,
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
    });
    await hb.tickNow();
    await hb.tickNow();
    expect(calls).toBe(2);
    hb.stop();
  });

  it('start + stop is idempotent', () => {
    const store = new MemoryTokenStore();
    const hb = createHeartbeat({
      baseUrl: 'https://x',
      fetchImpl: mockFetchJson(200, { success: true, data: {} }),
      store,
      licenseKey: 'KEY',
      fingerprint: 'fp',
      runtimeVersion: '1.0.0',
    });
    hb.start();
    hb.start(); // second start no-op
    hb.stop();
    hb.stop(); // second stop no-op — must not throw
  });
});

File: licensing/client/heartbeat_test.go
package client

import (
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
)

// Tests for SendOneHeartbeat. Coverage focuses on:
//
//  1. The request body shape matches the server contract (token, not
//     license_key + fingerprint + …) — earlier client versions sent the
//     wrong fields and the server returned 400 BadRequest for every
//     heartbeat. The first test locks the contract down.
//  2. Revocation auto-clear: a server response of LicenseRevoked /
//     LicenseSuspended causes the client to clear its local store so the
//     next guard / activate cycle prompts re-authentication.
//  3. Concurrency safety: a parallel Refresh that wrote a fresh token
//     between our Read and the typed-error reaction must NOT be
//     clobbered (CAS-style guard).

func TestSendOneHeartbeat_SendsTokenInRequestBody(t *testing.T) {
	var captured heartbeatRequest
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(body, &captured)
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"success": true,
			"data":    map[string]any{"ok": true},
		})
	}))
	defer srv.Close()

	store := NewMemoryTokenStore()
	_ = store.Write(StoredTokenState{Token: "LIC1.fake.token.value"})
	ok := SendOneHeartbeat(HeartbeatOptions{
		Store:     store,
		Path:      "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
	})
	if !ok {
		t.Fatal("heartbeat should have succeeded")
	}
	if captured.Token != "LIC1.fake.token.value" {
		t.Fatalf("server received wrong token: %q", captured.Token)
	}
}

func TestSendOneHeartbeat_NoStoreIsNoop(t *testing.T) {
	called := false
	srv := httptest.NewServer(http.HandlerFunc(func(http.ResponseWriter, *http.Request) {
		called = true
	}))
	defer srv.Close()
	ok := SendOneHeartbeat(HeartbeatOptions{
		Path:      "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
	})
	if !ok {
		t.Fatal("heartbeat with nil store should report ok")
	}
	if called {
		t.Fatal("heartbeat with nil store must not hit the network")
	}
}

func TestSendOneHeartbeat_EmptyStoreIsNoop(t *testing.T) {
	called := false
	srv := httptest.NewServer(http.HandlerFunc(func(http.ResponseWriter, *http.Request) {
		called = true
	}))
	defer srv.Close()
	ok := SendOneHeartbeat(HeartbeatOptions{
		Store:     NewMemoryTokenStore(), // empty
		Path:      "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
	})
	if !ok {
		t.Fatal("heartbeat with empty store should report ok")
	}
	if called {
		t.Fatal("heartbeat with empty store must not hit the network")
	}
}

func TestSendOneHeartbeat_ClearsGraceMarkerOnSuccess(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"success": true,
			"data":    map[string]any{"ok": true},
		})
	}))
	defer srv.Close()

	store := NewMemoryTokenStore()
	graceStart := int64(1_000_000)
	_ = store.Write(StoredTokenState{Token: "tok", GraceStartSec: &graceStart})

	if !SendOneHeartbeat(HeartbeatOptions{
		Store: store, Path: "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
	}) {
		t.Fatal("expected success")
	}
	final, _ := store.Read()
	if final.GraceStartSec != nil {
		t.Fatalf("grace marker should have been cleared: %v", final.GraceStartSec)
	}
	if final.Token != "tok" {
		t.Fatalf("token should be preserved: %q", final.Token)
	}
}

func TestSendOneHeartbeat_ClearsStoreOnLicenseRevoked(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusForbidden)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"success": false,
			"error": map[string]any{
				"code":    "LicenseRevoked",
				"message": "license is revoked",
			},
		})
	}))
	defer srv.Close()

	store := NewMemoryTokenStore()
	_ = store.Write(StoredTokenState{Token: "soon-to-be-cleared"})

	var capturedErr error
	ok := SendOneHeartbeat(HeartbeatOptions{
		Store: store, Path: "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
		OnError:   func(e error) { capturedErr = e },
	})
	if ok {
		t.Fatal("heartbeat should have failed on revoked")
	}
	if !errors.Is(capturedErr, ErrLicenseRevoked) {
		t.Fatalf("OnError should fire LicenseRevoked, got %v", capturedErr)
	}
	final, _ := store.Read()
	if final.Token != "" {
		t.Fatalf("store must be cleared after revoke; got %q", final.Token)
	}
}

func TestSendOneHeartbeat_ClearsStoreOnLicenseSuspended(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusForbidden)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"success": false,
			"error": map[string]any{
				"code":    "LicenseSuspended",
				"message": "license is suspended",
			},
		})
	}))
	defer srv.Close()

	store := NewMemoryTokenStore()
	_ = store.Write(StoredTokenState{Token: "soon-to-be-cleared"})

	if SendOneHeartbeat(HeartbeatOptions{
		Store: store, Path: "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
	}) {
		t.Fatal("heartbeat should have failed on suspended")
	}
	final, _ := store.Read()
	if final.Token != "" {
		t.Fatalf("store must be cleared after suspend; got %q", final.Token)
	}
}

func TestSendOneHeartbeat_DoesNotClearStoreOnTransientErrors(t *testing.T) {
	// IssuerUnreachable / RateLimited / network errors must NOT clear
	// the store — only revoked/suspended do, because those are
	// authoritative server-side signals.
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Retry-After", "30")
		w.WriteHeader(http.StatusTooManyRequests)
	}))
	defer srv.Close()

	store := NewMemoryTokenStore()
	_ = store.Write(StoredTokenState{Token: "must-survive"})

	var capturedErr error
	if SendOneHeartbeat(HeartbeatOptions{
		Store: store, Path: "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
		OnError:   func(e error) { capturedErr = e },
	}) {
		t.Fatal("heartbeat should have failed on rate-limit")
	}
	if !errors.Is(capturedErr, ErrRateLimited) {
		t.Fatalf("expected ErrRateLimited, got %v", capturedErr)
	}
	final, _ := store.Read()
	if final.Token != "must-survive" {
		t.Fatalf("store must be preserved on rate-limit; got %q", final.Token)
	}
}

func TestSendOneHeartbeat_CASGuardAgainstParallelRefresh(t *testing.T) {
	// Simulate a parallel refresh writing a NEW token between our Read
	// and the LicenseRevoked reaction. The guard MUST detect that the
	// stored token is no longer the one we heartbeated with and skip
	// the clear, so the parallel-refresh result isn't clobbered.
	store := NewMemoryTokenStore()
	originalToken := "original-token"
	_ = store.Write(StoredTokenState{Token: originalToken})

	// Wrap the store with one that swaps the token after the first Read
	// call (which the heartbeat performs to capture state.Token).
	racingStore := &racingTokenStore{
		inner:    store,
		swapAt:   1, // swap after the heartbeat's first read
		newToken: "fresh-token-from-parallel-refresh",
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusForbidden)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"success": false,
			"error": map[string]any{
				"code":    "LicenseRevoked",
				"message": "license is revoked",
			},
		})
	}))
	defer srv.Close()

	if SendOneHeartbeat(HeartbeatOptions{
		Store: racingStore, Path: "/heartbeat",
		Transport: TransportOptions{BaseURL: srv.URL},
	}) {
		t.Fatal("heartbeat should have failed on revoked")
	}
	final, _ := store.Read()
	if final.Token != "fresh-token-from-parallel-refresh" {
		t.Fatalf("CAS guard failed: parallel-refresh token was clobbered; got %q",
			final.Token)
	}
	if !strings.Contains(final.Token, "fresh") {
		t.Fatal("expected fresh token to survive")
	}
}

// racingTokenStore emulates a parallel Refresh that writes a fresh
// token between the heartbeat's two Read calls. After `swapAt` reads,
// it switches the underlying state to `newToken`.
type racingTokenStore struct {
	inner    *MemoryTokenStore
	newToken string
	mu       sync.Mutex
	reads    int
	swapAt   int
}

func (s *racingTokenStore) Read() (StoredTokenState, error) {
	s.mu.Lock()
	s.reads++
	doSwap := s.reads == s.swapAt
	s.mu.Unlock()
	state, err := s.inner.Read()
	if doSwap {
		_ = s.inner.Write(StoredTokenState{Token: s.newToken})
	}
	return state, err
}

func (s *racingTokenStore) Write(state StoredTokenState) error {
	return s.inner.Write(state)
}

func (s *racingTokenStore) Clear() error {
	return s.inner.Clear()
}
